utils/litemap/src/serde.rs

```diff
@@ -102,23 +102,27 @@ where
         M: MapAccess<'de>,
     {
         let mut map = LiteMap::with_capacity(access.size_hint().unwrap_or(0));
+        let mut out_of_order = Vec::new();
 
         // While there are entries remaining in the input, add them
         // into our map.
         while let Some((key, value)) = access.next_entry()? {
-            // Try to append it at the end, hoping for a sorted map.
-            // If not sorted, insert as usual.
-            // This allows for arbitrary maps (e.g. from user JSON)
-            // to be deserialized into LiteMap
-            // without impacting performance in the case of deserializing
-            // a serialized map that came from another LiteMap
+            // Try to append it at the end, hoping for a sorted map. Otherwise, collect
+            // out of order items and extend the map with them later. This way, we give
+            // the implementation an opportunity to avoid quadratic costs from calling
+            // insert() with out of order items.
+            // Handling ordered inputs first allows for arbitrary maps (e.g. from user JSON)
+            // to be deserialized into LiteMap without impacting performance in the case of
+            // deserializing a serialized map that came from another LiteMap.
             if let Some((key, value)) = map.try_append(key, value) {
-                // Note: this effectively selection sorts the map,
-                // which isn't efficient for large maps
-                map.insert(key, value);
+                out_of_order.push((key, value));
```

Comment:
I considered making this based on `StoreMut::lm_push` and `Store::lm_into_iter` but that required changing the adding the `StoreIntoIterator` bound to R.